internal/pkg/deploy/cloudformation/stack/env.go

```diff
@@ -243,31 +244,57 @@ func (e *EnvStackConfig) transformParameters(
 	// Remove or transform each of the current parameters.
 	var params []*cloudformation.Parameter
 	for k, p := range curr {
-		if transformed := transform(p, old[k]); transformed != nil {
-			params = append(params, transformed)
+		currP := p
+		for _, transform := range transformFunc {
+			currP = transform(currP, old[k])
+		}
+		if currP != nil {
+			params = append(params, currP)
 		}
 	}
 	return params, nil
 }
 
-// transformEnvControllerParameters transforms a parameter such that it uses its previous value if:
-// 1. The parameter exists in the old template.
-// 2. The parameter is env-controller managed.
-// Otherwise, it returns the parameter untouched.
-func transformEnvControllerParameters(new cloudformation.Parameter, old *cloudformation.Parameter) *cloudformation.Parameter {
-	if old == nil { // The ParamKey doesn't exist in the old stack, use the new value.
-		return &new
+// transformEnvControllerParameters transforms an env-controller managed parameter.
+// If the parameter exists in the old template, it uses its previous value. Otherwise, it returns its new default value.
+func transformEnvControllerParameters(new, old *cloudformation.Parameter) *cloudformation.Parameter {
+	if new == nil {
+		return nil
 	}
-
-	var (
-		isEnvControllerManaged = make(map[string]struct{})
-		exists                 = struct{}{}
-	)
+	isEnvControllerManaged := make(map[string]struct{})
 	for _, f := range template.AvailableEnvFeatures() {
-		isEnvControllerManaged[f] = exists
+		isEnvControllerManaged[f] = struct{}{}
 	}
 	if _, ok := isEnvControllerManaged[aws.StringValue(new.ParameterKey)]; !ok {
-		return &new
+		return new
+	}
+	if old == nil { // The EnvController-managed parameter doesn't exist in the old stack. Use the new value.
+		return new
+	}
+	return &cloudformation.Parameter{
+		ParameterKey: new.ParameterKey,
+
+		// Ideally, we would set `UsePreviousValue: true` unfortunately CodePipeline template config does not support it.
+		// https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline-cfn-artifacts.html#w2ab1c21c15c15
+		ParameterValue: old.ParameterValue,
```

Comment:
isn't this the same as just returning old?

Reply:
Ideally the code here should be
```
ParameterKey: new.ParameterKey
UsePreviousValue: true
```
With this "ideal" code, there is no assumption on `old`'s `ParameterKey` name. This `transform...` function doesn't assume that `new.ParameterKey == old.ParameterKey`.

We can't `UsePreviousValue` because of the said reason in comment, and instead use `ParameterValue`. If we start returning `old` here, we are effectively adding the assumption that `new.ParameterKey == old.ParameterKey`.

Reply:
i see - that assumption doesn't exist. but it _is_ always true that `new.ParameterKey == old.ParameterKey`, right? (based on `transformParameters`)

Reply:
plus, what would `UsePreviousValue` do if `old.ParameterKey != new.ParameterKey`? i think using that field would imply that they're equal... otherwise there might not be a previous value to use

Reply:
you are right! updated the code :meow_blob: